### PR TITLE
Fix  #2981 - Update Profile to match TableProfile

### LINKED ISSUE #2981: Update Profile to match TableProfile
**Feature**
Based on https://github.com/open-metadata/OpenMetadata/issues/1311

**Describe the task**
Update how we compute metrics to match the structure of TableProfile in the API

### PR DESCRIPTION
**Pull request by pmbrull**
### Describe your changes :

This PR fixes https://github.com/open-metadata/OpenMetadata/issues/2981 and closes https://github.com/open-metadata/OpenMetadata/issues/2980

Reviewing the API definitions, we noticed that our current approach to computing profiling data was not properly synced with the TableProfile definition.

Instead of adding yet-another-step and mapping the results artificially, we have done a refactor on how the profiling data gets computed. We are now relying on a single Profiler object, which is able to compute both table and column metrics and itself handles all the columns in the table.

Its results can then be safely parsed into TableProfile without any issues.

In terms of the JSON Schema, we modified a bit the types of the metrics to better represent the results and added a couple of metrics that were not present and we are already computing. We also changed a few names of the metrics. ⚠️ **This might have an impact on the UI as well** @Sachin-chaurasiya, @darth-coder00  ⚠️ However, we still need to work on the UI anyway for this EPIC so we could maybe live with it for a while.

We also added a method to manually name the metrics, which will make it easier to map with TableProfile specs.

An interesting addition is the function `add_props` in metrics/core. As we now perform the metrics instantiation inside the Profiler, we are passing the class definition as an input parameter. However, there are metrics that require specific attributes such as `bins` or `expression`. To have the Profiler as generic as possible and not keep checking specific metrics for individualities, `add_props` dynamically prepares a new class definition overriding the `__init__` method to add any required attribute we pass as `**kwargs`.

This means that we can run `new_hist = add_props(bins=5)(Metrics.HISTOGRAM.value)` and `new_hist` will be a completely independent and new class definition that replicates `Metrics.HISTOGRAM.value` but will assign `bins=5` to all its instantiations.

Happy to discuss!

Thanks

#
### Type of change :

- [ ] Bug fix
- [x] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: catalog-rest-service/src/main/resources/json/schema/entity/data/table.json [hunk 1]
```diff
@@ -274,57 +274,65 @@
...
-          "type": "number"
+          "type": "integer"
         },
-        "nullProportion": {
+        "nullRatio": {
           "description": "No.of null value proportion in columns.",
           "type": "number"
         },
-        "missingPercentage": {
+        "missingRatio": {
```
Comment (pmbrull) [inline catalog-rest-service/src/main/resources/json/schema/entity/data/table.json:301]: If these changes are an issue now, I can revert this and change the naming on the python side. We can tackle it later if needed

**Comment by pmbrull** [comment]
Ill revert the datatype changes and handle them separatedly as it is a handful

File: catalog-rest-service/src/main/resources/json/schema/entity/data/table.json [hunk 2]
```diff
@@ -274,57 +274,65 @@
           "description": "Column Name.",
           "type": "string"
         },
-        "valuesCount": {
+        "count": {
           "description": "Total count of the values in this column.",
-          "type": "number"
```
Comment (harshach) [inline catalog-rest-service/src/main/resources/json/schema/entity/data/table.json:279]: reason to keep number is the python generated code will make it int and the range for the int in python is lower compare to java

File: ingestion/setup.py [hunk 3]
```diff
@@ -141,6 +141,8 @@ def get_long_description():
     # sklearn integration
     "scikit-learn==1.0.2",
     "pandas==1.3.5",
+    # Metrics sample data gen
+    "numpy==1.22.2",
```
Comment (harshach) [inline ingestion/setup.py:145]: @pmbrull do we need this?

File: catalog-rest-service/src/main/resources/json/schema/entity/data/table.json (hunk 1 above)
Comment (harshach) [reply catalog-rest-service/src/main/resources/json/schema/entity/data/table.json:301]: lets not change it. UI can fix it.

**Comment by harshach** [comment]
@pmbrull overall LGTM. Left couple of comments. If we don't need numpy lets not include it

**Comment by harshach** [comment]
also we need to update the sample data

File: catalog-rest-service/src/main/resources/json/schema/entity/data/table.json [hunk 4]
```diff
@@ -320,10 +320,18 @@
...
         "min": {
           "description": "Minimum value in a column.",
-          "type": "number"
+          "type": ["number", "integer", "string"]
         },
         "max": {
           "description": "Maximum value in a column.",
+          "type": ["number", "integer", "string"]
+        },
+        "minLength": {
```
Comment (pmbrull) [inline catalog-rest-service/src/main/resources/json/schema/entity/data/table.json:329]: Hi @harshach, reverted all other changes and removed numpy dependency. The only change I'd like to stick here is: adding `minLength` and `maxLength` and updating the type for `min` and `max`. Looks like the tests are passing and this way I can compute it for numbers and strings

**Review by harshach — APPROVED** [review]

**Pull request merged**